File: foundations-jdbc-scala_2/src/scala/dev/typr/foundations/scala/DuckDbTypes.scala
package dev.typr.foundations.scala

import dev.typr.foundations.{DuckDbType, DuckDbTypes => JavaDuckDbTypes}

/** Scala-friendly DuckDbType instances that use Scala types instead of Java boxed types.
  * All types from dev.typr.foundations.DuckDbTypes are available here, with primitives and BigDecimal converted to Scala types.
  *
  * Extend this class to add your own custom types to a shared set of type definitions.
  */
class DuckDbTypes {
  // Primitives - convert Java boxed types to Scala native types
  val tinyint: DuckDbType[Byte] = JavaDuckDbTypes.tinyint.bimap(b => b, b => b)
  val smallint: DuckDbType[Short] = JavaDuckDbTypes.smallint.bimap(s => s, s => s)
  val integer: DuckDbType[Int] = JavaDuckDbTypes.integer.bimap(i => i, i => i)
  val bigint: DuckDbType[Long] = JavaDuckDbTypes.bigint.bimap(l => l, l => l)
  val float_ : DuckDbType[Float] = JavaDuckDbTypes.float_.bimap(f => f, f => f)
  val double_ : DuckDbType[Double] = JavaDuckDbTypes.double_.bimap(d => d, d => d)
  val boolean_ : DuckDbType[Boolean] = JavaDuckDbTypes.boolean_.bimap(b => b, b => b)
  val bool: DuckDbType[Boolean] = JavaDuckDbTypes.bool.bimap(b => b, b => b)

  // BigDecimal - convert Java BigDecimal to Scala BigDecimal
  val decimal: DuckDbType[BigDecimal] = JavaDuckDbTypes.decimal.bimap(jbd => BigDecimal(jbd), sbd => sbd.bigDecimal)
  val numeric: DuckDbType[BigDecimal] = JavaDuckDbTypes.numeric.bimap(jbd => BigDecimal(jbd), sbd => sbd.bigDecimal)

  // Array types - convert Java boxed arrays to Scala native arrays
  val tinyintArray: DuckDbType[Array[Byte]] = JavaDuckDbTypes.tinyintArray.bimap(
    arr => arr.map(_.byteValue()),
    arr => arr.map(java.lang.Byte.valueOf)
  )
  val smallintArray: DuckDbType[Array[Short]] = JavaDuckDbTypes.smallintArray.bimap(
    arr => arr.map(_.shortValue()),
    arr => arr.map(java.lang.Short.valueOf)
  )
  val integerArray: DuckDbType[Array[Int]] = JavaDuckDbTypes.integerArray.bimap(
    arr => arr.map(_.intValue()),
    arr => arr.map(java.lang.Integer.valueOf)
  )
  val bigintArray: DuckDbType[Array[Long]] = JavaDuckDbTypes.bigintArray.bimap(
    arr => arr.map(_.longValue()),
    arr => arr.map(java.lang.Long.valueOf)
  )
  val floatArray: DuckDbType[Array[Float]] = JavaDuckDbTypes.floatArray.bimap(
    arr => arr.map(_.floatValue()),
    arr => arr.map(java.lang.Float.valueOf)
  )
  val doubleArray: DuckDbType[Array[Double]] = JavaDuckDbTypes.doubleArray.bimap(
    arr => arr.map(_.doubleValue()),
    arr => arr.map(java.lang.Double.valueOf)
  )
  val booleanArray: DuckDbType[Array[Boolean]] = JavaDuckDbTypes.booleanArray.bimap(
    arr => arr.map(_.booleanValue()),
    arr => arr.map(java.lang.Boolean.valueOf)
  )
  val decimalArray: DuckDbType[Array[BigDecimal]] = JavaDuckDbTypes.decimalArray.bimap(
    arr => arr.map(BigDecimal(_)),
    arr => arr.map(_.bigDecimal)
  )

  // Forward all other types directly from Java
  val hugeint = JavaDuckDbTypes.hugeint
  val utinyint = JavaDuckDbTypes.utinyint
  val usmallint = JavaDuckDbTypes.usmallint
  val uinteger = JavaDuckDbTypes.uinteger
  val ubigint = JavaDuckDbTypes.ubigint
  val uhugeint = JavaDuckDbTypes.uhugeint
  val real = JavaDuckDbTypes.real
  val float4 = JavaDuckDbTypes.float4
  val float8 = JavaDuckDbTypes.float8
  val varchar = JavaDuckDbTypes.varchar
  val text = JavaDuckDbTypes.text
  val string = JavaDuckDbTypes.string
  val char_ = JavaDuckDbTypes.char_
  val bpchar = JavaDuckDbTypes.bpchar
  val blob = JavaDuckDbTypes.blob
  val bytea = JavaDuckDbTypes.bytea
  val binary = JavaDuckDbTypes.binary
  val varbinary = JavaDuckDbTypes.varbinary
  val bit = JavaDuckDbTypes.bit
  val bitstring = JavaDuckDbTypes.bitstring
  val date = JavaDuckDbTypes.date
  val time = JavaDuckDbTypes.time
  val timestamp = JavaDuckDbTypes.timestamp
  val datetime = JavaDuckDbTypes.datetime
  val timestamptz = JavaDuckDbTypes.timestamptz
  val timetz = JavaDuckDbTypes.timetz
  val timestamp_s = JavaDuckDbTypes.timestamp_s
  val timestamp_ms = JavaDuckDbTypes.timestamp_ms
  val timestamp_ns = JavaDuckDbTypes.timestamp_ns
  val interval = JavaDuckDbTypes.interval
  val uuid = JavaDuckDbTypes.uuid
  val json = JavaDuckDbTypes.json
  val hugeintArray = JavaDuckDbTypes.hugeintArray
  val utinyintArray = JavaDuckDbTypes.utinyintArray
  val usmallintArray = JavaDuckDbTypes.usmallintArray
  val uintegerArray = JavaDuckDbTypes.uintegerArray
  val ubigintArray = JavaDuckDbTypes.ubigintArray
  val varcharArray = JavaDuckDbTypes.varcharArray
  val blobArray = JavaDuckDbTypes.blobArray
  val dateArray = JavaDuckDbTypes.dateArray
  val timeArray = JavaDuckDbTypes.timeArray
  val timestampArray = JavaDuckDbTypes.timestampArray
  val timestamptzArray = JavaDuckDbTypes.timestamptzArray
  val intervalArray = JavaDuckDbTypes.intervalArray
  val uuidArray = JavaDuckDbTypes.uuidArray
  val jsonArray = JavaDuckDbTypes.jsonArray
  val listBoolean = JavaDuckDbTypes.listBoolean
  val listTinyint = JavaDuckDbTypes.listTinyint
  val listSmallint = JavaDuckDbTypes.listSmallint
  val listInteger = JavaDuckDbTypes.listInteger
  val listBigint = JavaDuckDbTypes.listBigint
  val listFloat = JavaDuckDbTypes.listFloat
  val listDouble = JavaDuckDbTypes.listDouble
  val listVarchar = JavaDuckDbTypes.listVarchar
  val listUuid = JavaDuckDbTypes.listUuid
  val listDate = JavaDuckDbTypes.listDate
  val listTime = JavaDuckDbTypes.listTime
  val listTimestamp = JavaDuckDbTypes.listTimestamp
  val listTimestamptz = JavaDuckDbTypes.listTimestamptz
  val listDecimal = JavaDuckDbTypes.listDecimal
  val listHugeint = JavaDuckDbTypes.listHugeint
  val listInterval = JavaDuckDbTypes.listInterval
  val unknown = JavaDuckDbTypes.unknown

  // Forward static methods
  def decimal(precision: Int, scale: Int): DuckDbType[java.math.BigDecimal] =
    JavaDuckDbTypes.decimal(precision, scale)

  def varchar(length: Int): DuckDbType[String] =
    JavaDuckDbTypes.varchar(length)

  def char_(length: Int): DuckDbType[String] =
    JavaDuckDbTypes.char_(length)

  def bit(length: Int): DuckDbType[String] =
    JavaDuckDbTypes.bit(length)

  def ofEnum[E <: Enum[E]](enumTypeName: String, fromString: java.util.function.Function[String, E]): DuckDbType[E] =
    JavaDuckDbTypes.ofEnum(enumTypeName, fromString)
}

object DuckDbTypes extends DuckDbTypes

File: foundations-jdbc-kotlin/src/kotlin/dev/typr/kotlinfoundations/DuckDbTypes.kt
package dev.typr.kotlinfoundations

import dev.typr.foundations.DuckDbType
import dev.typr.foundations.SqlFunction
import dev.typr.foundations.DuckDbTypes as JavaDuckDbTypes

/**
 * Kotlin-friendly DuckDbType instances that use Kotlin types instead of Java boxed types.
 * All types from dev.typr.foundations.DuckDbTypes are available here, with primitives converted to Kotlin types.
 *
 * Extend this class to add your own custom types to a shared set of type definitions.
 */
open class DuckDbTypes {
    // Primitives - convert Java boxed types to Kotlin native types
    open val tinyint: DuckDbType<Byte> = JavaDuckDbTypes.tinyint.bimap(SqlFunction { it }, { it })
    open val smallint: DuckDbType<Short> = JavaDuckDbTypes.smallint.bimap(SqlFunction { it }, { it })
    open val integer: DuckDbType<Int> = JavaDuckDbTypes.integer.bimap(SqlFunction { it }, { it })
    open val bigint: DuckDbType<Long> = JavaDuckDbTypes.bigint.bimap(SqlFunction { it }, { it })
    open val float_: DuckDbType<Float> = JavaDuckDbTypes.float_.bimap(SqlFunction { it }, { it })
    open val double_: DuckDbType<Double> = JavaDuckDbTypes.double_.bimap(SqlFunction { it }, { it })
    open val boolean_: DuckDbType<Boolean> = JavaDuckDbTypes.boolean_.bimap(SqlFunction { it }, { it })
    open val bool: DuckDbType<Boolean> = JavaDuckDbTypes.bool.bimap(SqlFunction { it }, { it })

    // Forward all other types directly from Java
    open val decimal = JavaDuckDbTypes.decimal
    open val numeric = JavaDuckDbTypes.numeric
    open val hugeint = JavaDuckDbTypes.hugeint
    open val utinyint = JavaDuckDbTypes.utinyint
    open val usmallint = JavaDuckDbTypes.usmallint
    open val uinteger = JavaDuckDbTypes.uinteger
    open val ubigint = JavaDuckDbTypes.ubigint
    open val uhugeint = JavaDuckDbTypes.uhugeint
    open val real = JavaDuckDbTypes.real
    open val float4 = JavaDuckDbTypes.float4
    open val float8 = JavaDuckDbTypes.float8
    open val varchar = JavaDuckDbTypes.varchar
    open val text = JavaDuckDbTypes.text
    open val string = JavaDuckDbTypes.string
    open val char_ = JavaDuckDbTypes.char_
    open val bpchar = JavaDuckDbTypes.bpchar
    open val blob = JavaDuckDbTypes.blob
    open val bytea = JavaDuckDbTypes.bytea
    open val binary = JavaDuckDbTypes.binary
    open val varbinary = JavaDuckDbTypes.varbinary
    open val bit = JavaDuckDbTypes.bit
    open val bitstring = JavaDuckDbTypes.bitstring
    open val date = JavaDuckDbTypes.date
    open val time = JavaDuckDbTypes.time
    open val timestamp = JavaDuckDbTypes.timestamp
    open val datetime = JavaDuckDbTypes.datetime
    open val timestamptz = JavaDuckDbTypes.timestamptz
    open val timetz = JavaDuckDbTypes.timetz
    open val timestamp_s = JavaDuckDbTypes.timestamp_s
    open val timestamp_ms = JavaDuckDbTypes.timestamp_ms
    open val timestamp_ns = JavaDuckDbTypes.timestamp_ns
    open val interval = JavaDuckDbTypes.interval
    open val uuid = JavaDuckDbTypes.uuid
    open val json = JavaDuckDbTypes.json
    open val tinyintArray = JavaDuckDbTypes.tinyintArray
    open val smallintArray = JavaDuckDbTypes.smallintArray
    open val integerArray = JavaDuckDbTypes.integerArray
    open val bigintArray = JavaDuckDbTypes.bigintArray
    open val hugeintArray = JavaDuckDbTypes.hugeintArray
    open val utinyintArray = JavaDuckDbTypes.utinyintArray
    open val usmallintArray = JavaDuckDbTypes.usmallintArray
    open val uintegerArray = JavaDuckDbTypes.uintegerArray
    open val ubigintArray = JavaDuckDbTypes.ubigintArray
    open val floatArray = JavaDuckDbTypes.floatArray
    open val doubleArray = JavaDuckDbTypes.doubleArray
    open val decimalArray = JavaDuckDbTypes.decimalArray
    open val booleanArray = JavaDuckDbTypes.booleanArray
    open val varcharArray = JavaDuckDbTypes.varcharArray
    open val blobArray = JavaDuckDbTypes.blobArray
    open val dateArray = JavaDuckDbTypes.dateArray
    open val timeArray = JavaDuckDbTypes.timeArray
    open val timestampArray = JavaDuckDbTypes.timestampArray
    open val timestamptzArray = JavaDuckDbTypes.timestamptzArray
    open val intervalArray = JavaDuckDbTypes.intervalArray
    open val uuidArray = JavaDuckDbTypes.uuidArray
    open val jsonArray = JavaDuckDbTypes.jsonArray
    open val listBoolean = JavaDuckDbTypes.listBoolean
    open val listTinyint = JavaDuckDbTypes.listTinyint
    open val listSmallint = JavaDuckDbTypes.listSmallint
    open val listInteger = JavaDuckDbTypes.listInteger
    open val listBigint = JavaDuckDbTypes.listBigint
    open val listFloat = JavaDuckDbTypes.listFloat
    open val listDouble = JavaDuckDbTypes.listDouble
    open val listVarchar = JavaDuckDbTypes.listVarchar
    open val listUuid = JavaDuckDbTypes.listUuid
    open val listDate = JavaDuckDbTypes.listDate
    open val listTime = JavaDuckDbTypes.listTime
    open val listTimestamp = JavaDuckDbTypes.listTimestamp
    open val listTimestamptz = JavaDuckDbTypes.listTimestamptz
    open val listDecimal = JavaDuckDbTypes.listDecimal
    open val listHugeint = JavaDuckDbTypes.listHugeint
    open val listInterval = JavaDuckDbTypes.listInterval
    open val unknown = JavaDuckDbTypes.unknown

    // Parameterized methods
    open fun decimal(precision: Int, scale: Int): DuckDbType<java.math.BigDecimal> =
        JavaDuckDbTypes.decimal(precision, scale)

    open fun varchar(length: Int): DuckDbType<String> = JavaDuckDbTypes.varchar(length)

    open fun char_(length: Int): DuckDbType<String> = JavaDuckDbTypes.char_(length)

    open fun bit(length: Int): DuckDbType<String> = JavaDuckDbTypes.bit(length)

    open fun <E : Enum<E>> ofEnum(enumTypeName: String, fromString: java.util.function.Function<String, E>): DuckDbType<E> =
        JavaDuckDbTypes.ofEnum(enumTypeName, fromString)

    companion object : DuckDbTypes()
}
